### PR TITLE
Imagecropper

### PR DESCRIPTION
**Pull request by GedMarc**
#6077 

Use JDK 8 native Path for referencing, important for JDK 11 and up particularly with modules enabled

### CODE REVIEW

**Comment by melloware** [comment]
@GedMarc looks like you have 3 checkstyle violations.
```
[ERROR] /home/travis/build/primefaces/primefaces/src/main/java/org/primefaces/component/imagecropper/ImageCropperRenderer.java:228: 'else' child has incorrect indentation level 28, expected level should be 24. [Indentation]
[ERROR] /home/travis/build/primefaces/primefaces/src/main/java/org/primefaces/component/imagecropper/ImageCropperRenderer.java:228:1: File contains tab characters (this is the first instance). [FileTabCharacter]
[ERROR] /home/travis/build/primefaces/primefaces/src/main/java/org/primefaces/component/imagecropper/ImageCropperRenderer.java:229: 'else' child has incorrect indentation level 28, expected level should be 24. [Indentation]
```

File: src/main/java/org/primefaces/component/imagecropper/ImageCropperRenderer.java [hunk 1]
```diff
@@ -220,15 +224,9 @@ else if (imageObject instanceof StreamedContent) {
...
-                        if (!(webRoot.endsWith("\\") || webRoot.endsWith("/")) &&
-                                    !(imagePath.startsWith("\\") || imagePath.startsWith("/"))) {
-                            fileSeparator = "/";
-                        }
-
-                        File file = new File(webRoot + fileSeparator + imagePath);
-                        inputStream = new FileInputStream(file);
+                        String webRoot = externalContext.getRealPath(imagePath);
+	                    Path locationPath  = Paths.get(webRoot);
+	                    inputStream = Files.newInputStream(locationPath, StandardOpenOption.READ);
```
Comment (melloware) [inline src/main/java/org/primefaces/component/imagecropper/ImageCropperRenderer.java:229]: CheckStyle

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by GedMarc** [comment]
Hmm so I'm trying to isolate here, this looks to be a 2 parts here so I'm going to need a bit of guidance - 

This fix is working for me right now, i'll try clean it all up and see what repercussions  i have caused myself

The Paths should take care of everything linux/mac/ and updated. The NIO library has huge updates relating to file management, e.g. using FileStream here may be replaceable with FileChannel (MacOS gets usually a 60% file read increase, i'll find the link somewhere), 

For me in particular, JLink native imaging uses the JRT filesystem... which is readonly, so i have another problem where the uploaded files are trying to be stored in the classpath on a read only module native image xD i can route to a static location easy enough but can't seem to specify where to save the files

* How do I add a custom location to be specified for the uploaded files, in the showcase when i push the button it goes to /demo/images/....

:D :D

**Comment by melloware** [comment]
I think it has to be relative to the web application root, and that is why we had to add the OWASP Path Traversal check: 
https://github.com/primefaces/primefaces/issues/3268

So there is no way to specify some other file location i think it has to be under the web app.  But I did not design this component  it has been like this for years.

**Comment by GedMarc** [comment]
Oh dear.... 

We do either a static location on a server more designed for static content and reverse proxy a location into there
or save the data in a table to be served by a servlet at another location. 
Saving or updating the actual web application with file data, hmmm...

Can you lead me to the widget that is implemented? Is this something that we are open to? Retrieving the data in the same manner as file upload into a stream rather than saving it down to the file system?

For the showcase can always display the crop image in a base 64 data image saving on some disk space for the showcase server?

On the wildfly bug for 6.2, another solution may have been to go this route, and specify ${jboss.temp.directory} as the target directory to make sure it is retrievable on the same context url. It is possible to do the same behavior specifying file saving locations?

**Comment by melloware** [comment]
So if you have 9.0-SNAPSHOT running locally and grab 9.0-SNAPSHOT Showcase from master you can see recently these 2 ticket were implemented: 
https://github.com/primefaces/primefaces/issues/12
https://github.com/primefaces/primefaces/issues/4837

And 2 new Showcase examples were added for ImageCropper `Dynamic` and `Combined With FileUpload`.

image

But yes the showcase is always saving the cropped images in \primefaces-showcase\src\main\webapp\resources\demo\images\crop

**Pull request closed without merging**